### ISSUE TITLE
Devise::FailureApp setting response status 401 gets ignored by Rails 5 alpha version code and instead returning 200 code

### ISSUE THREAD

**Issue by jiggneshhgohel** [rails/rails#22439]
I am working on making my Rails 4.2.0/Ruby 2.2.1 **API-only app** ready for Rails 5/Ruby 2.2.3 and for which I am using the Rails 5 edge version. My Gemfile.lock has following entry for Rails, Devise, Rack etc.

```
GIT
  remote: git://github.com/twalpole/devise.git
  revision: 85700fba52891bb73e607b9414dc27e246721c6b
  branch: rails5
  specs:
    devise (3.5.2)
      bcrypt (~> 3.0)
      orm_adapter (~> 0.1)
      railties (>= 3.2.6, < 5)
      responders
      thread_safe (~> 0.1)
      warden (~> 1.2.3)

GIT
  remote: https://github.com/rack/rack.git
  revision: c393176b0edf3e5d06cabbb6eb9d9c7a07b2afa7
  specs:
    rack (2.0.0.alpha)
      json

GIT
  remote: https://github.com/rails/arel.git
  revision: 3c429c5d86e9e2201c2a35d934ca6a8911c18e69
  specs:
    arel (7.0.0.alpha)

GIT
  remote: https://github.com/rails/rails.git
  revision: 5217db2a7acb80b18475709018088535bdec6d30
  specs:
    actionmailer (5.0.0.alpha)
      actionpack (= 5.0.0.alpha)
      actionview (= 5.0.0.alpha)
      activejob (= 5.0.0.alpha)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 1.0, >= 1.0.5)
    actionpack (5.0.0.alpha)
      actionview (= 5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      rack (~> 2.x)
      rack-test (~> 0.6.3)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    activejob (5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      globalid (>= 0.3.0)
    activemodel (5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      builder (~> 3.1)
    activerecord (5.0.0.alpha)
      activemodel (= 5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      arel (= 7.0.0.alpha)
    activesupport (5.0.0.alpha)
      concurrent-ruby (~> 1.0)
      i18n (~> 0.7)
      json (~> 1.7, >= 1.7.7)
      method_source
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    rails (5.0.0.alpha)
      actionmailer (= 5.0.0.alpha)
      actionpack (= 5.0.0.alpha)
      actionview (= 5.0.0.alpha)
      activejob (= 5.0.0.alpha)
      activemodel (= 5.0.0.alpha)
      activerecord (= 5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      bundler (>= 1.3.0, < 2.0)
      railties (= 5.0.0.alpha)
      sprockets-rails (>= 2.0.0)
    railties (5.0.0.alpha)
      actionpack (= 5.0.0.alpha)
      activesupport (= 5.0.0.alpha)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)

```

I have a controller spec for my customized Devise SessionsController in my app:

```
  context "POST #create" do
    context "responds with 401 error" do
      it "when the user trying to sign-in with the given email doesn't exists" do
        params[:user].merge!(email:  'inexistent_user@example.com')
        post :create, params: params

        expect(response).to have_http_status(401)
        expect(json_response["error"]).to include("Invalid email or password")
      end
  end
end
```

That spec with Rails 4.2.0/Ruby 2.2.1 is passing. However running that spec against Rails 5 version code it is failing with error saying:

```
     Failure/Error: expect(response).to have_http_status(401)
       expected the response to have status code 401 but it was 200
```

The response status code 401 set by Devise  [Devise::FailureApp](https://github.com/twalpole/devise/blob/85700fba52891bb73e607b9414dc27e246721c6b/lib/devise/failure_app.rb#L21) is getting ignored. I traced the execution flow as highlighted in trace below:

```
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/failure_app.rb:21:in `call'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/failure_app.rb:21:in `call'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/delegator.rb:5:in `call'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/test_helpers.rb:113:in `_process_unauthenticated'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/test_helpers.rb:91:in `_catch_warden'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/devise-85700fba5289/lib/devise/test_helpers.rb:19:in `process'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/rails-5217db2a7acb/actionpack/lib/action_controller/test_case.rb:610:in `process_with_kwargs'
/home/jignesh/.rvm/gems/ruby-2.2.3@myapp-on-rails-5/bundler/gems/rails-5217db2a7acb/actionpack/lib/action_controller/test_case.rb:371:in `post'
```

The trace then further proceeds up to class method [`self.action(name)` in `ActionController::Metal`](
https://github.com/rails/rails/blob/5217db2a7acb80b18475709018088535bdec6d30/actionpack/lib/action_controller/metal.rb#L252) which invokes class method `self.make_response!(request)`.

```
    def self.make_response!(request)
      ActionDispatch::Response.create.tap do |res|
        res.request = request
      end
    end

    # Returns a Rack endpoint for the given action name.
    def self.action(name)
      if middleware_stack.any?
        middleware_stack.build(name) do |env|
          req = ActionDispatch::Request.new(env)
          res = make_response! req
          new.dispatch(name, req, res)
        end
      else
        lambda { |env|
          req = ActionDispatch::Request.new(env)
===>  res = make_response! req
          new.dispatch(name, req, res)
        }
      end
    end
```

`self.make_response!(request)` invokes [`ActionDispatch::Response.create`](https://github.com/rails/rails/blob/5217db2a7acb80b18475709018088535bdec6d30/actionpack/lib/action_dispatch/http/response.rb#L133) class method which has the default status set to 200.

I think this is the root cause behind my spec failure. Can anybody please confirm if this a bug or no?

Thanks.

**Comment by nickgnd** [comment]
HI @jiggneshhgohel,  I've the same problem with Rails 5 (API only) and Ruby 2.2.3 like you.
I use **Warden** to authenticate the user (that is included in Devise).

When I test the controller with Rspec **I'm expecting 401 status for unauthenticated user but I receive 200 as response status**. **It happens only in Test environment**, if I try in Development env it returns 401 like expected (see https://github.com/nickgnd/example_rails_api/blob/test/rails5alpha/README.md).
Instead, **the response body is set correctly** both in test and development environment (I didn't try in production env).

I don't understand if my problem is in  [ActionDispatch::Response.create](https://github.com/rails/rails/blob/5217db2a7acb80b18475709018088535bdec6d30/actionpack/lib/action_dispatch/http/response.rb#L133) as you said or in the [Class](https://github.com/nickgnd/example_rails_api/blob/test/rails5alpha/spec/support/warden.rb) that I use to stub Warden middleware for testing (see note). 

With Rails 4.2 all works fine.

Some weeks ago i push a clean repo to **reproduce this bug step by step** if someone want to try.
https://github.com/nickgnd/example_rails_api/tree/test/rails5alpha
Branch: rails5alpha
The master branch is with rails 4.2.3.


Note: 
* the helper that I use to test Warden is based on Devise helpers. 
source: http://kentaroimai.com/articles/1-controller-test-helpers-for-warden

**Comment by jiggneshhgohel** [comment]
@nickgnd thanks for confirming that I am not alone facing this problem.

Regarding your point that it happens in Test environment only I can recollect that in the test env logs I could see 401 returned but while asserting response code in spec example it is found to be 200.

BTW I am using Warden test helpers but those are included only for specs tagged with `type: :request` and the problem I am encountering is in Controller spec example. So I don't think this has anything to do with Warden.

Below I have given an extract from my `spec/rails_helper.rb`

```

RSpec.configure do |config|
  # Remove this line if you're not using ActiveRecord or ActiveRecord fixtures
  config.fixture_path = "#{::Rails.root}/spec/fixtures"

  # If you're not using ActiveRecord, or you'd prefer not to run each of your
  # examples within a transaction, remove the following line or assign false
  # instead of true.

  # References:
  #   https://relishapp.com/rspec/rspec-rails/docs/transactions
  #   http://stackoverflow.com/questions/21493970/databasecleaner-rspec-what-is-the-correct-configuration
  #   http://devblog.avdi.org/2012/08/31/configuring-database_cleaner-with-rails-rspec-capybara-and-selenium/
  config.use_transactional_fixtures = false

  # https://github.com/DatabaseCleaner/database_cleaner#rspec-example
  config.before(:suite) do
    DatabaseCleaner.strategy = :transaction
    DatabaseCleaner.clean_with(:truncation)
  end

  config.before(:suite) do
    begin
      DatabaseCleaner.start
      FactoryGirl.lint
    ensure
      DatabaseCleaner.clean
    end
  end

  config.around(:each) do |example|
    DatabaseCleaner.cleaning do
      example.run
    end
  end

  config.include FactoryGirl::Syntax::Methods

  # RSpec Rails can automatically mix in different behaviours to your tests
  # based on their file location, for example enabling you to call `get` and
  # `post` in specs under `spec/controllers`.
  #
  # You can disable this behaviour by removing the line below, and instead
  # explicitly tag your specs with their type, e.g.:
  #
  #     RSpec.describe UsersController, :type => :controller do
  #       # ...
  #     end
  #
  # The different available types are documented in the features, such as in
  # https://relishapp.com/rspec/rspec-rails/docs
  config.infer_spec_type_from_file_location!

  config.include Devise::TestHelpers, type: :controller

  # Reference: https://github.com/plataformatec/devise/wiki/How-To:-sign-in-user-in-Request-type-specs-%28specs-tagged-with-type:-:request%29
  config.include Warden::Test::Helpers, type: :request

end
```

**Comment by nickgnd** [comment]
Hi @jiggneshhgohel, in the previous comment I expressed myself badly, I corrected it.

Like your case, the error comes out only in the specs when I assert the response status.

**Comment by nickgnd** [comment]
I tried to debug my spec and I found a different behavior compared to your, but that always responds with 401.

If I put a **breakpoint** in my [UnautheniticatedController](https://github.com/nickgnd/example_rails_api/blob/test/rails5alpha/app/controllers/unauthenticated_controller.rb) (that is used from Warden to handle unauthorized request) and I start to debug from here, I find that **the response status change after the method get(action, *args) is called**. 
This method is located in [actionpack/lib/action_controller/test_case.rb](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/test_case.rb#L362).

```ruby
# Simulate a GET request with the given parameters.
# ...
def get(action, *args)
        res = process_with_kwargs("GET", action, *args)
        cookies.update res.cookies
        res
end
```
The method receive this arguments:
* action: index
* args: {:format=>:json, :method=>"GET"}

and return **res**:  #<ActionDispatch::TestResponse:0x007fdafd952b00> object.
=> res.status = 200

If I inspect the res object (```res.inspect```) I find some trace of the original 401 status:
(the complete trace [here](http://pastebin.com/M8hdVXVZ))

```bash
...
@buf=["{\"errors\":[\"Unauthorized Request\"]}"], @closed=false, @str_body=nil>,
@status=401, @cv=#<MonitorMixin::ConditionVariable:0x007fdafe2b8320 
@monitor=#<ActionDispatch::Response:0x007fdafe2b8488 ...>, @cond=#<Thread::ConditionVariable:0x007fdafe2b82f8>>, @committed=true, @sending=true, @sent=true, @cache_control={}, @request=#<ActionDispatch::Request:0x007fdafe2b8550 @env={...}, @filtered_parameters=nil, @filtered_env=nil, @filtered_path=nil, @protocol=nil, @port=nil, @method=nil, @request_method=nil, @remote_ip=nil, @original_fullpath=nil, @fullpath=nil, @ip=nil>>, @_routes=nil, @_action_name="respond", 
@_response_body=["{\"errors\":[\"Unauthorized Request\"]}"]>, "warden"=>Warden::Proxy:70289272574460 @config={:default_scope=>:default, :scope_defaults=>{}, :default_strategies=>{:_all=>[:authentication_token]}, 
:intercept_401=>true, :failure_app=>UnauthenticatedController}, "HTTP_X_USER_EMAIL"=>nil, "HTTP_X_AUTH_TOKEN"=>nil, "HTTP_COOKIE"=>"", "PATH_INFO"=>"/unauthenticated", "action_dispatch.request.path_parameters"=>{:format=>"json", :controller=>"customers", :action=>"index"}, ...
```

The method **get(action, *args)** follow this other custom method in [spec/support/warden.rb](https://github.com/nickgnd/example_rails_api/blob/test/rails5alpha/spec/support/warden.rb#L36) that I use to stub Warden Middleware.
```ruby
# Catch warden continuations and handle like the middleware would.
# Returns nil when interrupted, otherwise the normal result of the block.
  def _catch_warden(&block)
        result = catch(:warden, &block)

        if result.is_a?(Hash) && !warden.custom_failure? && !@controller.send(:performed?)
          result[:action] ||= :unauthenticated

          env = @controller.request.env
          env['PATH_INFO'] = "/#{result[:action]}"
          env['warden.options'] = result
          Warden::Manager._run_callbacks(:before_failure, env, result)

          status, headers, body = warden.config[:failure_app].call(env).to_a
          
         @controller.send :render, status: status, plain: body,
            content_type: headers['Content-Type'], location: headers['Location']

          nil
        else
          result
       end
end
```
The argument passed to the method ```:render``` are:
* status: 401
* headers: {"Access-Control-Allow-Origin"=>"*", "Access-Control-Allow-Methods"=>"GET,POST,PUT,DELETE,OPTIONS", "Access-Control-Allow-Headers"=>"Content-Type,Accept,X-User-Email,X-Auth-Token", "Content-Type"=>"application/json; charset=utf-8"}
(byebug)
* body: #<ActionDispatch::Response::RackBody:0x007fdaff133b20>

(see debug below fore more info)

### Debug step
```bash
$ bin/rspec spec/controllers/customers_controller_spec.rb

[9, 18] in /Users/Nico/RoR/example_rails_api/app/controllers/unauthenticated_controller.rb
    9:     self.status        = :unauthorized
   10:     self.content_type  = "application/json"
   11:     self.response_body = { errors: ["Unauthorized Request"] }.to_json
   12:     self.headers["Access-Control-Allow-Origin"]  = CORS_ALLOW_ORIGIN
   13:     self.headers["Access-Control-Allow-Methods"] = CORS_ALLOW_METHODS
   14:     self.headers["Access-Control-Allow-Headers"] = CORS_ALLOW_HEADERS
   15:     byebug
=> 16:   end
   17: 
   18: end
(byebug) status
401
(byebug) next

[188, 197] in /Users/Nico/.rvm/gems/ruby-2.2.3/bundler/gems/rails-e37b470a6675/actionpack/lib/action_controller/metal.rb
   188: 
   189:     def dispatch(name, request, response) #:nodoc:
   190:       set_request!(request)
   191:       set_response!(response)
   192:       process(name)
=> 193:       request.commit_flash
   194:       to_a
   195:     end
   196: 
   197:     def set_response!(response) # :nodoc:
(byebug) response.status
401
(byebug) next

[189, 198] in /Users/Nico/.rvm/gems/ruby-2.2.3/bundler/gems/rails-e37b470a6675/actionpack/lib/action_controller/metal.rb
   189:     def dispatch(name, request, response) #:nodoc:
   190:       set_request!(request)
   191:       set_response!(response)
   192:       process(name)
   193:       request.commit_flash
=> 194:       to_a
   195:     end
   196: 
   197:     def set_response!(response) # :nodoc:
   198:       @_response = response
(byebug) to_a
401
{"Access-Control-Allow-Origin"=>"*", "Access-Control-Allow-Methods"=>"GET,POST,PUT,DELETE,OPTIONS", "Access-Control-Allow-Headers"=>"Content-Type,Accept,X-User-Email,X-Auth-Token", "Content-Type"=>"application/json; charset=utf-8"}
#<ActionDispatch::Response::RackBody:0x007fdafaf34d28>
(byebug) next

[45, 54] in /Users/Nico/RoR/example_rails_api/spec/support/warden.rb
   45:           env['PATH_INFO'] = "/#{result[:action]}"
   46:           env['warden.options'] = result
   47:           Warden::Manager._run_callbacks(:before_failure, env, result)
   48: 
   49:           status, headers, body = warden.config[:failure_app].call(env).to_a
=> 50:           @controller.send :render, status: status, plain: body,
   51:             content_type: headers['Content-Type'], location: headers['Location']
   52: 
   53:           nil
   54:         else
(byebug) status
401
(byebug) next

[359, 368] in /Users/Nico/.rvm/gems/ruby-2.2.3/bundler/gems/rails-e37b470a6675/actionpack/lib/action_controller/test_case.rb
   359:       #
   360:       # Note that the request method is not verified. The different methods are
   361:       # available to make the tests more expressive.
   362:       def get(action, *args)
   363:         res = process_with_kwargs("GET", action, *args)
=> 364:         cookies.update res.cookies
   365:         res
   366:       end
   367: 
   368:       # Simulate a POST request with the given parameters and set/volley the response.
(byebug) action
index
(byebug) args
{:format=>:json, :method=>"GET"}
(byebug) res
#<ActionDispatch::TestResponse:0x007fdafd8d8f58>
(byebug) res.status
200
(byebug) 
```

**Comment by jiggneshhgohel** [comment]
@nickgnd I have observed that Devise correctly sets the response code as 401. After that is done and as the execution moves ahead the status gets set to 200. Your debug process involves debugging a GET request and mine involves a POST request so I guess the findings differs however I tried to debug the chain at its root and as part of this I have found what I have mentioned in my first comment.

Let's see if we can hear something on this from the core contributors list. They should be able to shed more light on what might be causing this.

Thanks.

**Comment by maclover7** [comment]
Please open up a ticket on Devise's repo for any issues with Devise.

**Comment by dadarek** [comment]
@nickgnd or @jiggneshhgohel - were you able to solve your problem in any way? I am running into the same issue, but I am moving from rails 4.0 to rails 4.1.

**Comment by nickgnd** [comment]
@dadarek Yep. Simply I switched from `controller` tests to [`integration` tests](http://api.rubyonrails.org/v4.2.5/classes/ActionDispatch/IntegrationTest.html). 
For me it was very easy (I worked on a API Rails app, so I usually test response status and body).

In my case the bug was in the stub of `Warden` middleware and I I haven't been able to solve it. 
Differently from the `controller` tests, `integration` tests load all the middleware (`Warden` included), so you don't need to stub them, and this has solved the problem :smile: 

A standard `controller` test change from this: 
```ruby
class ArticlesControllerTest < ActionController::TestCase
  test "should get index" do
    get :index
    assert_response :success
  end
end
```

to this:
```ruby
class ArticlesTest  < ActionDispatch::IntegrationTest
   test "should get index" do
    get '/index'
    assert_response :success
  end
end
```

I hope it helps
